crates/bevy_reflect/src/struct_trait.rs

```diff
@@ -40,25 +46,170 @@ pub trait Struct: Reflect {
     /// `&mut dyn Reflect`.
     fn field_mut(&mut self, name: &str) -> Option<&mut dyn Reflect>;
 
+    /// Returns an iterator over the values of the struct's fields.
+    fn iter_fields(&self) -> FieldIter;
+
+    /// Clones the struct into a [`DynamicStruct`].
+    fn clone_dynamic(&self) -> DynamicStruct;
+
+    /// Returns the number of fields in the struct.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the [`TypeInfo`] returned by [`Reflect::get_type_info()`] is not
+    /// [`TypeInfo::Struct`]— which should almost never be the case. Notable exceptions
+    /// include [`DynamicStruct`] which uses its own implementation of this method to
+    /// prevent the panic.
+    fn field_len(&self) -> usize {
+        if let TypeInfo::Struct(info) = self.get_type_info() {
+            info.field_len()
+        } else {
+            panic!(
+                "tuple `{:?}` is not `TypeInfo::Struct`",
+                std::any::type_name::<Self>()
+            );
+        }
+    }
+
+    /// Returns the name of the field with index `index`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the [`TypeInfo`] returned by [`Reflect::get_type_info()`] is not
+    /// [`TypeInfo::Struct`]— which should almost never be the case. Notable exceptions
+    /// include [`DynamicStruct`] which uses its own implementation of this method to
+    /// prevent the panic.
+    fn name_at(&self, index: usize) -> Option<&str> {
+        if let TypeInfo::Struct(info) = self.get_type_info() {
+            info.field_at(index).map(|field| field.name().borrow())
+        } else {
+            panic!(
+                "tuple `{:?}` is not `TypeInfo::Struct`",
+                std::any::type_name::<Self>()
+            );
+        }
+    }
+
     /// Returns a reference to the value of the field with index `index` as a
     /// `&dyn Reflect`.
-    fn field_at(&self, index: usize) -> Option<&dyn Reflect>;
+    ///
+    /// # Panics
+    ///
+    /// Panics if the [`TypeInfo`] returned by [`Reflect::get_type_info()`] is not
+    /// [`TypeInfo::Struct`]— which should almost never be the case. Notable exceptions
+    /// include [`DynamicStruct`] which uses its own implementation of this method to
+    /// prevent the panic.
+    fn field_at(&self, index: usize) -> Option<&dyn Reflect> {
+        if let TypeInfo::Struct(info) = self.get_type_info() {
```

Comment:
> If we're going to go in this direction, I want to see benchmarks backing this up for the following cases (comparing the old and new approaches for the methods mentioned above):
> 
> * A relatively "average" generic and non-generic struct
> * A struct with a huge number of fields, with access to fields that would be "at and near the end" of the match statements (to properly establish O(1) vs O(n) behavior for each impl).

Yep, good call. I'll run some benchmarks to see how much faster/slower it is.

Reply:
Okay, ran a simple benchmark. This tests getting the first (0), middle (50), and last (100) field of a generic struct with 100 fields. Looks like using `TypeInfo` is about 20–60x slower. The non-generic case is about the same but just _slightly_ better (only 10–50x slower).
<img width="855" alt="reflect_comparison" src="https://user-images.githubusercontent.com/49806985/172742053-6d8724f1-6562-453f-aa37-02139d507af8.png">

This means we'll probably want to revert and remove the default implementations.

Reply:
Going to second reverting this. Optimizations like the ones in #4081 strictly rely on the simple match's simplicity and speed. We're going to need this for high performance reflect-based animation.

Reply:
Reverted :)

Reply:
Before closing the book on perf analysis: did you make sure that you "pre-computed" the static type info before running the bench?

Can you post the benchmarks somewhere?

Reply:
I'm also a bit curious about the reverse: `field_at` is almost certainly going to be better as a match. But mapping field names to fields might actually benefit from using the static hashmap, especially for large structs / long field names.

Reply:
> Before closing the book on perf analysis: did you make sure that you "pre-computed" the static type info before running the bench?

Yeah, I create a function called `prepare` that I run before the tests which simply registers the test struct, which should theoretically initialize the `TypeInfo`.

> Can you post the benchmarks somewhere?

Sure thing. Here's the link to the repo: [bevy_reflect_benchmark](https://github.com/MrGVSV/bevy_reflect_benchmark);

Hopefully I did things right, I'm not too experienced with writing benchmarks haha. I also only tested the largest struct, but I can do all of them if that's preferred.

Reply:
Here's the results of the small struct (3 fields):
<img width="879" alt="reflect_comparison_small" src="https://user-images.githubusercontent.com/49806985/172746106-f6243e44-4d0c-43df-9870-52576e59277e.png">

Reply:
> I'm also a bit curious about the reverse: field_at is almost certainly going to be better as a match. But mapping field names to fields might actually benefit from using the static hashmap, especially for large structs / long field names.

ahhh i missed that you already did this test. my bad!

Reply:
And yeah I think the evidence is clearly in favor of the revert. I think we can call this resolved.

Reply:
I agree that the implementation was suboptimal, but I don't think it's inherent to being a separate struct.

I wanted to address this:

> New TypeInfo (Non-Generic)
>
>    call get_type_info() (an atomic pointer access because we're assuming we've already constructed it)
>    call field_at, which will use TypeInfo's lookup table to return the NamedField reference
>    call name(), which will grab the field name pointer
>    call field(name), which will do a match on string names (which to my knowledge won't be optimized to use lookup tables, so will be O(n) string comparisons), then return the field reference

On x64, atomic pointer-sized loads are the same instruction as non-atomic pointer-sized loads. Only writes differ.

To resolve the other performance concerns, the API can be adjusted to the following (approximate):

```rust
// on TypeInfo:
unsafe fn field_ptr(&self, value_ptr: *const dyn Reflect, idx: usize) -> Option<*const dyn Reflect> {
    let offset = self.field_offsets.get(idx)?; // field_offsets: Vec<usize>
    let field_reflect_metadata = self.field_metadata.get(idx)?; // field_reflect_metadata: Vec<core::ptr::DynMetadata>
    Some(core::ptr::from_raw_parts(value_ptr.cast::<()>().add(offset), field_reflect_metadata))
}
fn field_ref<'r>(&self, value: &'r dyn Reflect, field_idx: usize) -> Option<&'r dyn Reflect> {
    let value_ptr: *const dyn Reflect = value;
    self.field_ptr(value_ptr, field_idx).map(|ptr| &*)
}
```

`core::ptr::DynMetadata` is not currently stabilized, but it's possible to emulate the behaviour for this case using other unsafe constructs.